src/main/java/org/gridsuite/shortcircuit/server/computation/service/AbstractWorkerService.java

```diff
@@ -0,0 +1,240 @@
+            if (future != null) {
+                future.cancel(true);  // cancel computation in progress
+            }
+            cleanResultsAndPublishCancel(cancelContext.resultUuid(), cancelContext.receiver());
+        } finally {
+            lockRunAndCancel.unlock();
+        }
+    }
+
+    protected abstract AbstractResultContext<R> fromMessage(Message<String> message);
+
+    protected boolean resultCanBeSaved(S result) {
+        return result != null;
+    }
+
+    public Consumer<Message<String>> consumeRun() {
+        return message -> {
+            AbstractResultContext<R> resultContext = fromMessage(message);
+            try {
+                AtomicReference<Long> startTime = new AtomicReference<>();
+                startTime.set(System.nanoTime());
+
+                Network network = getNetwork(resultContext.getRunContext().getNetworkUuid(),
+                        resultContext.getRunContext().getVariantId());
+                S result = run(network, resultContext.getRunContext(), resultContext.getResultUuid());
+
+                long nanoTime = System.nanoTime();
+                LOGGER.info("Just run in {}s", TimeUnit.NANOSECONDS.toSeconds(nanoTime - startTime.getAndSet(nanoTime)));
+
+                if (resultCanBeSaved(result)) {
+                    observer.observe("results.save", resultContext.getRunContext(), () -> saveResult(network, resultContext, result));
+
+                    long finalNanoTime = System.nanoTime();
+                    LOGGER.info("Stored in {}s", TimeUnit.NANOSECONDS.toSeconds(finalNanoTime - startTime.getAndSet(finalNanoTime)));
+
+                    sendResultMessage(resultContext, result);
+                    LOGGER.info("{} complete (resultUuid='{}')", getComputationType(), resultContext.getResultUuid());
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
+                if (!(e instanceof CancellationException)) {
+                    LOGGER.error(NotificationService.getFailedMessage(getComputationType()), e);
+                    publishFail(resultContext, e.getMessage());
+                    resultService.delete(resultContext.getResultUuid());
+                }
+            } finally {
+                futures.remove(resultContext.getResultUuid());
+                cancelComputationRequests.remove(resultContext.getResultUuid());
+            }
+        };
+    }
+
+    public Consumer<Message<String>> consumeCancel() {
+        return message -> cancelAsync(CancelContext.fromMessage(message));
+    }
+
+    protected abstract void saveResult(Network network, AbstractResultContext<R> resultContext, S result);
+
+    protected abstract void sendResultMessage(AbstractResultContext<R> resultContext, S result);
+
+    protected abstract void publishFail(AbstractResultContext<R> resultContext, String message);
```

Comment:
In my opinion, we should implement the default behaviour to avoid the implementation in others children classes

Reply:
not yet done